### ISSUE TITLE
TryOptionAsync.ToTry should be be fail when None

### ISSUE THREAD

**Issue by sanderploegsma** [louthy/language-ext#623]
Given a `TryOptionAsync` that is `None`, I would expect `.ToTry()` to become `Fail`, since the `TryOptionAsync` is not successful. However, this test fails:

``` c#
[Fact]
public async void TestToTry()
{
    var result = TryOptionAsync(() => Task.FromResult<Option<string>>(None)).ToTry();
    Assert.True(await result.IsFail());
}
```

It seems that the conversion to `TryAsync` returns `new Result<A>(default (A))`, which in most cases evaluates to `null`. This seems very counterintuitive.

How do I convert a `TryOptionAsync` to a `TryAsync`, making sure that the `None` state is handled as a failure?

**Comment by bender2k14** [comment]
I think this behavior comes from [`OptionalResult<A>.ToResult()`](https://github.com/louthy/language-ext/blob/master/LanguageExt.Core/DataTypes/Result/OptionalResult.cs#L209-L214), which is implemented as
```
internal Result<A> ToResult() =>
    IsFaulted
        ? new Result<A>(Exception)
        : Value.IsSome
            ? new Result<A>(Value.Value)
            : new Result<A>(default(A));
```

My first expectation is that Language.Ext would not define a mapping from `OptionalResult<A>` to `Result<A>`.  Given that it has defined such a mapping, my next expectation is that it would throw an exception (like it does when [calling Reduce on an empty list](https://github.com/louthy/language-ext/issues/583)).  I did not expect that it would create a default value for `A` and give the impression that the status is nominal.

@louthy, is the implementation of `OptionalResult<A>.ToResult()` as you intended?

**Comment by louthy** [comment]
This is a clear bug.  `ToTry` should probably return `TryAsync<Option<A>` here, otherwise the behaviour isn't obvious.

**Comment by louthy** [comment]
This is fixed in `v3.1.15`.

`TryOption.ToTry` and `TryOptionAsync.ToTry` now return `Try<Option<A>>` and `TryAsync<Option<A>>` respectively.

There's also a new `ToTry` extension for both that takes a `None` delegate for mapping `None -> A` to remove the `Option`
